packages/common/src/converter/payload-converter.ts

```diff
@@ -319,6 +333,60 @@ export class SearchAttributePayloadConverter implements PayloadConverter {
 
 export const searchAttributePayloadConverter = new SearchAttributePayloadConverter();
 
+export class TypedSearchAttributePayloadConverter implements PayloadConverter {
+  jsonConverter = new JsonPayloadConverter();
+
+  public toPayload<T>(typedSearchAttribute: T): Payload {
+    if (!isTypedSearchAttribute(typedSearchAttribute)) {
+      throw new ValueError('Invalid typed search attribute');
+    }
+
+    const type = typedSearchAttribute[0];
+    let value = typedSearchAttribute[1];
+    if (type === 'DATETIME') {
+      // Convert Date to ISO string
+      value = (value as Date).toISOString();
+    }
+
+    const payload = this.jsonConverter.toPayload(value);
+    if (payload === undefined) {
+      throw new ValueError('Could not convert typed search attribute to payload');
+    }
+    // Note: this shouldn't be the case but the compiler complains without this check.
+    if (payload.metadata === undefined || payload.metadata === null) {
+      throw new ValueError('Missing payload metadata');
+    }
+    // Add encoded type of search attribute to metatdata
+    payload.metadata['type'] = encode(toMetadataType(type));
+    return payload;
+  }
+
+  public fromPayload<T>(payload: Payload): T {
+    if (payload.metadata === undefined || payload.metadata === null) {
+      throw new ValueError('Missing payload metadata');
+    }
+
+    // TODO(thomas): type casting undefined values is not clear to caller
+
+    let value = this.jsonConverter.fromPayload(payload);
+    // If no 'type' metadata field or no given value, we skip.
+    if (payload.metadata.type === undefined || value === undefined) {
+      return undefined as T;
+    }
+    const type = toSearchAttributeType(decode(payload.metadata.type));
```

Comment:
In that case, better make the return type of this function `T | undefined`, to make it clear that the caller has the responsibility of handling that possibility.

Reply:
I can't change the type of this function because it breaks the interface . The `PayloadConverter` interface definition seems very restrictive, the way it's typed makes it difficult to get much type-safety in the `toPayload` and `fromPayload` definitions.

In any case, I left a comment noting why this was the case and what the suggested usage is. Not clear to me how to make this better.